### PR TITLE
Workflows and dependencies update + python 3.12 support

### PR DESCRIPTION
**Pull request by Smartappli**
# Related Issue or bug

Info about Issue or bug

Closes #[issue number that will be closed through this PR]

# Describe the changes you've made

A clear and concise description of what you have done to successfully close your assigned issue. Any new files? or anything you feel to let us know!

# Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. 

# Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)

A clear and concise description of it.

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

# Screenshots

 Original           | Updated
 :--------------------: |:--------------------:
 **original screenshot**  | <b>updated screenshot </b> |

### CODE REVIEW

**Comment by Smartappli** [comment]
https://github.com/Smartappli/pycaret/actions/runs/9656014246/job/26632822533

**Comment by Smartappli** [comment]
@celestinoxp @Yard1 PR ready to review :)

**Comment by celestinoxp** [comment]
@Smartappli can you add badge in readme.md with python 3.12 ?

**Comment by Smartappli** [comment]
> @Smartappli can you add badge in readme.md with python 3.12 ?

Done

File: requirements.txt [hunk 1]
```diff
@@ -7,7 +7,7 @@ pandas>=2.2.0 #latest sktime makes tests fail if pandas 2.1
 jinja2>=3
 scipy>=1.6.1,<=1.11.4 #to fix later (https://github.com/mljar/mljar-supervised/issues/691)
 joblib>=1.2.0,<1.4  # joblib<1.2.0 is vulnerable to Arbitrary Code Execution (https://github.com/advisories/GHSA-6hrg-qmvc-2xh8)  # joblib 1.4.0 breaks PyCaret
-scikit-learn>1.4.0 #scikit-learn 1.4.0 have a bug (https://github.com/scikit-learn/scikit-learn/pull/28241)
+scikit-learn>1.4.0,<1.5.0 #scikit-learn 1.4.0 have a bug (https://github.com/scikit-learn/scikit-learn/pull/28241)
```
Comment (celestinoxp) [inline requirements.txt:10]: why doesn't pycaret support scikit-learn 1.5 ?

Comment (Smartappli) [reply requirements.txt]: Reverted thx for your comment

**Comment by celestinoxp** [comment]
@Yard1 can you review this PR ?

**Review by celestinoxp — APPROVED** [review]
@Yard1 @moezali1 @tvdboom @ngupta23 
can anyone review this PR ?

**Comment by celestinoxp** [comment]
distutils will no longer available by default in python 3.12. Pycaret have a reference here pycaret/utils/_dependencies.py. The only way to support it its install setuptools package, the maaybe a good idea add setuptools  to requirements.txt with confition if installation is with python 3.12... what you think about it ?
references: https://peps.python.org/pep-0632/

maybe in requirements.txt use setuptools; python_version == '3.12' ?

**Comment by Smartappli** [comment]
> distutils will no longer available by default in python 3.12. Pycaret have a reference here pycaret/utils/_dependencies.py. The only way to support it its install setuptools package, the maaybe a good idea add setuptools to requirements.txt with confition if installation is with python 3.12... what you think about it ? references: https://peps.python.org/pep-0632/
> 
> maybe in requirements.txt use setuptools; python_version == '3.12' ?
> 
> @Yard1 what you think about it?

@celestinoxp Thank you for your advising.
Change done

**Review by celestinoxp — APPROVED** [review]

**Comment by celestinoxp** [comment]
With this update, maybe a good idea bump version to pycaret 3.5 ?

**Review by Yard1 — APPROVED** [review]
Changes look good, will merge and release if CI passes.

**Pull request merged**